ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,647 @@
+choose as few such slots as possible such that no sparse window remain.
+
+Moreover, each window needs a different number of slots (because some windows
+may be more sparse than others). Because some windows may require multiple
+empty slots, this is a /set multi-covering problem/.
+
+Lastly, we can only toggle an empty slot once (recall that it's a /zero-one/
+variable), so this is a /set multi-cover with multiplicity constraints
+problem/.
+
+After some searching, I found this reference
+
+  * Qiang-Sheng Hua, Yuexuan Wang, Dongxiao Yu, Francis C. M. Lau: Dynamic
+    programming based algorithms for set multicover and multiset multicover
+    problems. Theor. Comput. Sci. 411(26-28): 2467-2474 (2010)
+
+which includes the claim "[...] we give the first known exact algorithm for the
+MSMC or the SMC with multiplicity constraints problem".
+
+It /may/ be the case that our problem is particularly degenerate, and so easier
+to solve. In particular, our variables are binary, their coefficients in the
+constraints are binary, and in fact our constraints form a banded matrix. Only
+the lower bound of each constraint can be greater than 1. But I have not yet
+recognized how to do it.
+
+The closest I have come to a solution is an incremental algorithm that toggles
+one empty slot at a time. I /think/ the completeness of this algorithm requires
+that the next toggle is one of the empty slots that occurs in the most number
+of sparse windows. However, I have not proved that. Moreover, even if it is
+true, a counterexample proves that choosing an arbitrary such slot is not
+sufficient:
+
+>     A   B   C
+> -----   -----
+> 01110111011101110
+>     -----   -----
+
+In this example, only slots A, B, and C occur in two sparse windows, so those
+are the candidates for the next toggle by my conjecture. If we suppose that
+each window is only missing one active slot (eg need 4 actives per 5 slots),
+then toggling B would require three toggles in total, whereas toggling A and C
+would solve the problem with just two toggles.
+
+-}
+
+-- | A 'ChainSchedule' that satisfies 'checkHonestChain'
+--
+-- The distribution this function samples from is not simple to describe. It
+-- begins by drawing a sample of length 'Len' from the Bernoulli process
+-- induced by the active slot coefficient 'Asc'. (IE 'Len' many i.i.d. samples
+-- from @Uniform(f)@). Then it slides an @s-d@ window along that leader
+-- schedule, toggling as many empty slots as necessary to ensure every such
+-- window has at least @k+1@ active slots, ie Enriched Praos Chain
+-- Growth---HOWEVER, it is a difficult computation to add a /strictly minimal/
+-- number of active slots to the output of the Bernoulli process (it's an
+-- integer programming optimization problem as far as we can tell). We have
+-- settled for a relatively simple algorithm that approaches the minimal number
+-- of toggles (TODO calculate how tight the bounds are).
```

Comment:
Could this note explain or point to a description of what the simple algorithm is?

Reply:
Done in f0e51da8